### DISCUSSION TITLE
Additional Form/Field Layout Options

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#9161] [Feature Requests]
Let's discuss the following Collection/Field Form Layout Options:

* Show field label on left (field-level) **Show visual option**
  * Mostly Interface
  * Equal Sizing
  * Mostly Label
  * Fit to Interface
* Add top/bottom border
  * If two borders overlap, combine
* Hide label (alias only?)
  * Could use field right-click when hidden?
* Note Above + Note Below

Could we add a "Form Preview" to the Data Model > Collection page? Or perhaps find a way to have it represent the form UI better? What if we made the form editable from the Item Page too... perhaps admins get additional Field Label Context Menu options?

Interface Options to add:

* Textarea: character count (or connect to validation?)
* Text Input Copy Value
* Text Input Prefix/Suffix (subdued text before/after)
* New Toggle UI for boolean
* New Group style (see design)
* Add clear button to dropdown (if enabled)
* Add "Note" alias interface

![Item Detail Form UI](https://user-images.githubusercontent.com/522079/138915444-6e0dddf7-80df-4482-be2f-a87f680cbca9.png)

**Comment by Jonesus** [comment]
While the different mockups all do look quite visually pleasing, it has been studied that aligning input labels on top of their respective fields results in fastest time to fill forms and least eye movement (and thus strain) on the user. Therefore I would wish to at least have the option to force labels to be on top of their respective inputs :pray: 
https://www.uxmatters.com/mt/archives/2006/07/label-placement-in-forms.php

**Reply by benhaynes** [reply]
An excellent point! 👍  Yes, what we have now will stay available, as it is a very standard field layout flow. What I am proposing are adding a few additional field-level options for adding greater control over the design.

A lot of what a admin might choose depends on other factors, such as how much (if any) helper text (notes) is required, and any necessary field grouping.

**Comment by hermesalvesbr** [comment]
One suggestion would be a data search field, autocomplete style with action to see the data after the click.

Like looking for a product, to make a sale. Or suggest another way to do this?

**Comment by OzzyW80** [comment]
I agree Ben, having flexibility in this would really up the developer/admin experience. What we have experienced is that when a workspace have massive amount of data fields the labels on top make each form unecessarily long. What could have been resolved in one screen takes several scrolls. On the UX side: to get an overview of all fields in one screen or as few scrolls as possible may be overriding the label on top advantages in some cases.

One idea would be to also allow admins to align the labels left or right. Forms with right aligned labels on the side may look a little disorganized but are really pleasing for the end user. 

And lastly to be able to add more than 2 fields on one row would be a massive improvement. But I can imagine it may be complicated to develop.

@benhaynes

**Reply by rijkvanzanten** [reply]
As part of this bigger discussion, we're actually thinking through how we could make this work based on a six-column grid rather than a two-column grid. That'll allow you to do 3/3 like now, or put 3 fields side by side in a 2/2/2 configuration. (I don't think we'd want to allow for 1/1/1/1/1/1, as interfaces don't really scale down that far)

**Reply by benhaynes** [reply]
Exactly. The more I think about this, the more I want to use the same layout system as in Insights. We have a grid (6 columns) and you can create whatever you want. Interfaces could have a minimum width/height required that they can't shrink below. The main thing to consider if we go with this direction is responsiveness and vertical form flow (many interfaces _grow_ in height based on usage). 🤔

**Reply by benhaynes** [reply]
For sizing, I could imagine checkbox interfaces fitting in a 2 column (or even 1 column) grid, but a WYSIWYG having a 3-column min.

**Reply by benhaynes** [reply]
Clearly, this also opens up the door to injecting Panels (from Insights) and other readonly components into the form too. 👍

**Reply by rijkvanzanten** [reply]
> The more I think about this, the more I want to use the same layout system as in Insights. We have a grid (6 columns) and you can create whatever you want

That being said, insights actually has an infinitely growing canvas, rather than a 6-column layout 🤔

**Reply by benhaynes** [reply]
Correct, we would have a "grid of dots" that is a static 6 across (could add more for an "extended form", I guess) and X auto-growing tall. If we really want to lean into this approach, we could even offer a "grid" area above Layouts (we discussed allowing charts, etc up there too).

**Reply by rijkvanzanten** [reply]
We'll have to investigate how to make that approach work responsively, and especially with growing interfaces (like you mentioned). It'll be difficult to "push things down" if you're working on a 2-dimensional grid like that 🤔

**Reply by OzzyW80** [reply]
Such great mockups @benhaynes @rijkvanzanten 

Some additional thoughts from me (some reasoning by comparison, sorry about it but kind of useful when discussing alternatives in design):

- Yes the idea of insights-like-flexible creator mode is great
- Regarding the 6 columns I benchmarked vs our accounting system and they have 4 columns (and they don't have a sidebar so I could imagine 6 could appear crowded but worth a try). The thing in their case is one column could contain 3 fields in their case. I see the greatness in this because you can then both (1) provide as many valuable fields/data as possible in one screen and (2) group fields contextually in columns.
- It would be great to be allowed to hide label. Like in the case of Toggl plan here below they write To-From to the left, then it could be enough with just two date fields. All fields may not need a label to be understandable.
- It would be a great improvement to have placeholder in date field. Like "From date" and "To date" or just "From" / "To"
- The toggles in the mockup looks amazing
- It would be great to be able to adjust label on the left to the middle of the form field if there is no help text available
- Totally agree with @JonL1 -> what is important to solve for is to reduce the scrolling. To achieve this I see: More columns / Multiple field in column / Horizontal form submenu / Smaller form fields but this can be done with custom css / Accordion as already implemented (such great group type imo)
- Yes I agree with @Jonesus and I believe this is because actually user stop focusing on the labels and start focusing on the placeholders. However I can totally see the benefits of the toggl plan version below here. You are able to fit alot of fields into the same screen with placement of label to the left. Since each row might have multiple fields the form labels almost becomes like a headline instead. Just opinions

And finally to summarize what I think is important is the overview-perspective of the user. She want to simply just have the right fields in front of her without too much scroll if they can be grouped contextually -great. Over time the labels become secondary for navigation, so if the labels consume too much inventory it is not so good over time.

I add three references here. Their UI is actually unmodern. But the UX is actually great and they are very popular by users. 

1) Fortnox the accounting system - 400 k non-tech users in a country of 10 million inhabitants: 4 columns, multiple fields per column so system designer can adapt the size of the field to the content, they also actually have smaller size of form fields but since we have the custom css never mind about that

2) A basic invoicing system - this is such a popular and self explanatory system, and their forms are completely messed up but since the fields are placed contextually they become self explanatory - they also apply the scroll-as-little-as-possible idea to their system. And it is crazy but the traditional grey background with white form fields (just like you created with the mailing address above Ben) is so much better than white background to white form field for light mode. It is so interesting. My wife uses this system, she is totally non-technical, but she never had a question about using the system. No onboarding needed whatsoever. Everything was self explanatory to her. That level of UX is definitely something to strive for!

3) Toggl plan - I think they as well as Notion and others really cracked the "have one field cateogry per row". You can get so many fields into one modal or screen with that method.

<img width="1440" alt="Skärmavbild 2022-02-23 kl  22 41 41" src="https://user-images.githubusercontent.com/96046228/155413381-f24f8111-0ced-4aba-b63c-a278f2ea210a.png">

<img width="996" alt="Skärmavbild 2022-02-23 kl  22 43 59" src="https://user-images.githubusercontent.com/96046228/155413746-480bd134-93e6-4722-9e4a-54811afad462.png">

<img width="560" alt="Skärmavbild 2022-02-23 kl  22 43 09" src="https://user-images.githubusercontent.com/96046228/155413597-6f189e9d-df31-42c8-aeb9-3a4bf6948cda.png">

**Reply by OzzyW80** [reply]
> Clearly, this also opens up the door to injecting Panels (from Insights) and other readonly components into the form too. 👍

Combining forms with lists would be amazing. Imagine you have this small form on top with say 2 rows. Press create, the item is displayed in the list below.

**Reply by benhaynes** [reply]
Really great insight and feedback, @OzzyW80 — all really great points. The big question with these more complex forms will be if/how they stay responsive. We can always update fields from 1/2 to 1/1 (etc) at certain breakpoints... but depending on how you design your form, that wrapping might get... funky (counter-intuitive).

Really excited to add these additional options into our form design system... it'll be interesting to see what a good UI for configuring this will look like. It's representational now (just little boxes), but it might be nice to make it more "design in place" (WYSIWYG).

Lot's to cover! 😐

**Reply by OzzyW80** [reply]
I can imagine responsiveness could become really complicated. I would however expect you to just apply your everyday magic to that challenge :)

**Reply by OzzyW80** [reply]
@benhaynes I posted below, actually forms does not have to be scrollable vertically.

**Comment by JonL1** [comment]
I like the mockups. Specially the new grouping style.

I would like to throw in an idea for a new type of grouping for the detail view: multi step wizard layout.
The ability of grouping the whole form or parts of it in a multi-step wizard layout.

Some forms tend to get big and to reduce scrolling this could be an alternative plus it gives a sense of flow for users when completing one.

**Reply by benhaynes** [reply]
Yeah! We have a `group` type now, and this has a few variants: raw, group, accordion. We could definitely look into adding "steps" and "tabs" for these purposes. 👍

**Reply by Pachat** [reply]
Steps are great for long forms. 
It would be great also if we could choose to send the step's data to the db when we leave, blur or unload a step.
That brings more security on long forms and gives the ability to fill it with different sessions on different devices.
_Long forms examples: insurance declaration file or application for admission to university_

**Reply by Pachat** [reply]
A nice to have is that when going back to the form, the user is redirected to the step where he/she left.

**Comment by mdocter** [comment]
Here's my two cents...
I would like to see forms having the following nesting. This is based on the form builder in Microsoft Dynamics 365 CE.

Form
- Tabs
- - Tab(s)
- - - Row(s)
- - - - Section(s)
- - - - - Field(s)

A Form can have a fluid/full width, or fixed and centered width. 

Tabs can be shown one at a time (normal tabbed behavior) or vertically stacked below each other. If you go for the last option, at least provide a drop-down at the top of the form to scroll/jump to a tab. In the vertical stacked layout a tab acts like a collapsible section which by default can open expanded or collapsed (on form load). Hopefully you provide both layout types. 
Microsoft currently uses the vertically stacked tabs, but the users I spoke to where always more delighted by the tabbed interface. Probably less visual noise and only see the tab contents you're interested in.

Row elements in a tab are vertically stacked and take up the full width of the tab, but inner sections are horizontally aligned (and wrapped). 
A row can have one to X columns. 

Each inner section should define the number of columns of the row it occupies. The section also defines if the field labels within are placed above the field or to the left/right of the field.

Fields are placed within a column of the section. Each column has a vertical layout. So fields are stacked. A field can span multiple section columns. So for example, a three column section can contain a field which spans three columns and therefore has the same width as the section. On the next line, there a three fields which are all one section column wide. So each field occupies one third of the section width.

This nesting/grouping, when realized with flexbox, can make the form layout adjust to each screen size. From wide screens to mobile with the right wrapping of the container elements. 

Nice to have:
- Drag and drop form editor;
- Role based / multiple forms per collection;
- Quick create forms with only required fields or similar. Useful from actions above subgrids. Can be presented in a dialog or in the right side slide-out pane;
- Inner-/Subgrids to show lists of related records in the form sections.

Hope this provides some valuable input.
Great job guys! Still really impressed by Directus. Keep up the good work!

**Reply by benhaynes** [reply]
Thank you, @mdocter!!

> A Form can have a fluid/full width, or fixed and centered width.

I'd want to still include the option to left align a max-width form. But I like this... essentially moves the `fill` we have for fields to the form level.

As for tabs, we already have support for Grouping (eg: Accordion) and are planning to add Tabs soon.

The nesting (rows and sections) will be up to @rijkvanzanten — but below is a more flexible 6-column grid I've been thinking about. One thing that would make a big difference is to do form setup (drag and drop) when you can actually see the interfaces... so it's a bit more WYSIWYG (see dated mockup below). Not sure if this would be in Settings or on the actual item detail page.

![Settings-Data-Model-Collection](https://user-images.githubusercontent.com/522079/164909118-952c706c-9511-4b83-acd8-6b35c1b8c093.png)

**Reply by AustinPhillipTaylor** [reply]
@benhaynes Looks really nice! Do you have any objections to a 12-column grid? 12 columns is used in a lot of interfaces as it is easily divisible into 1/6, 1/4, 1/3, 1/2. For our purposes, a 1/12 width column may be a little too thin, however, I believe we could restrict the smallest width to span at least 2 columns, if we think it might be an issue. The only difference to the 6-column layout is the availability of the 1/4 width, but I think that's a very important breakpoint.

**Reply by VaZark** [reply]
@benhaynes have you given thought to offering a paid or cloud-only nocode dashboard builder for public websites ? (Like Elementor)

Between this, displays, interfaces, global theming and extensions... you guys already have all you need. you’ll need to do the tedious task of extracting out the design logic for no immediate benefit tho. The App can be treated as a compiled build of the proposed port.

**Reply by AustinPhillipTaylor** [reply]
I believe something like that has been discussed. However, unless I'm misremembering, I think that's something that's been determined be be better suited for an extension. When you get right into it, it's much more complex of a task than it appears on the surface.

**Reply by benhaynes** [reply]
> Do you have any objections to a 12-column grid?

No, but I don't think any meaningful fields could fit in 1/12... or even 1/4, so I have distilled that down to 6-columns to keep things a bit more organized. That said, we could just have the interfaces set min-widths so you can still have 4-up checkboxes... but not 4-up maps. Open to discussing!

> nocode dashboard builder for public websites

Yeah! We're pushing the current no-code app forward, and will also keep adding more ways to share data, forms, and dashboards publically. ❤️

**Reply by AustinPhillipTaylor** [reply]
I think it could go either way. At the end of the day, it depends on what you prefer to be handled by the fields, versus the internals of the interface. 

If we want more to be handled at the field level, I could see a use-case for 1/12 to 1/4 for things like check boxes/toggle switches, radio buttons, thumbnail images (that'd be an interesting interface to explore), color inputs, or really any interface that handles its interaction by opening a secondary window/pop-up. 

On the other hand, it does make sense to delegate a lot of the size handling to the interface, and just handle general placement using the field-level columns. In that case, you're right, 6 columns is a good idea because it allows 1/3 without breaking 1/2.

**Reply by benhaynes** [reply]
Adding this here in case it's relevant. It's the flip side of this new Data Model mockup.
![Settings-Data-Model-Collection](https://user-images.githubusercontent.com/522079/174836141-0ea3247d-ee84-44e1-8520-d0a977f9e464.png)

**Reply by OzzyW80** [reply]
Incredible plans @benhaynes 

Basic thing that could add value: On collection there is a Notes field. This type of field would also be valuable to have in the "Field slideover page" so we can add comments to the purpose of each field. Personally think it would be better to have it as a WYSIWYG. We are currently adding comments on the fields in an XLS for documentation purposes.

**Reply by benhaynes** [reply]
Absolutely — we'll just need to be clear on the difference between our (Directus) field notes... and the schema's column comments. Also, our notes do already support Markdown, so you can do simple formatting and links. 👍

**Comment by hugopvl** [comment]
Hi Guys, 
I actually have some very long forms, and if we could have a 6-column layout it would be perfect, will we ever have this option available?

**Reply by rijkvanzanten** [reply]
How would 6 columns scale down to mobile view, or would it jump from 1 to 6 once the viewport gets big enough? 🤔

**Reply by rijkvanzanten** [reply]
Just a little worried about fields being 1cm wide otherwise 👀

**Reply by mdocter** [reply]
I can imagine 1 on mobile, 3 on tablet, 6 > tablet screens. So if it can be made configurable per viewport width...

**Reply by rijkvanzanten** [reply]
> So if it can be made configurable per viewport width...

That adds a ton of configuration overhead though, cause then you'd have to configure the form multiple times 😬  Maybe we could make the 3 an "automated" 3 based on effectively cutting out the right half of the 6 and moving them down a row? We'll have to think through how to handle cases where an interface goes across the middle though 🤔

**Reply by mdocter** [reply]
Sorry, I thought this was the thread with my original form/tabs/sections proposal... 🤗
Just ignore my previous comment.

**Reply by OzzyW80** [reply]
@rijkvanzanten I am also working on a massive database with long forms. If the tablet navigation could also be available as selectable it would improve UX. If we could add more field columns it would also be great attaching an example here

If the form groups are setup as horizontal maybe the fields does not have to collapse on mobile but be scrollable horizontally

In this example they have 4 groups that can be up to 3 fields each totaling to 12 fields in a row. 

I spend +100 hrs/week in Directus and am loving it. Thank you for bringing it to the world

**Reply by OzzyW80** [reply]
<img width="1632" alt="Screenshot 2023-01-11 at 11 59 26 PM" src="https://user-images.githubusercontent.com/96046228/211936325-3f1951a8-070e-4c3f-8a4f-87e9f641186f.png">
<img width="507" alt="Screenshot 2023-01-11 at 11 59 57 PM" src="https://user-images.githubusercontent.com/96046228/211936410-502db040-2a47-4c93-a3c3-60da86c34d27.png">

**Reply by rijkvanzanten** [reply]
Overflow scroll could work, but is often a little hostile in UX 🤔 It's different for something that's intentionally read-only though like Insights

**Reply by LasseRosenow** [reply]
What about this mockup by the founder, he posted some time ago: https://github.com/directus/directus/discussions/9161#discussioncomment-2620819

It seems it would address this request. Or is it discarded?

**Reply by rijkvanzanten** [reply]
> It seems it would address this request.

It does and it doesn't! I'm fully down with having a 6-column grid, the main thing we haven't solved for yet though is how it scales back down for mobile / smaller viewports 🙂

**Reply by LasseRosenow** [reply]
> It does and it doesn't!

This is so poetic :o 🙂

Just to add my 2 cents I guess there would not be a "perfect solution" that would totally satisfy all screen sizes the way everyone would like it too, but maybe even if the solution is not perfect, like for example needing to scroll on mobile, or having them wrap on mobile, it could still be a huge benefit for many customers, who only want to use the desktop environment anyways.

My guess is that probably most people who actively use directus, would be using a desktop viewport and thus could benefit a lot by this new feature. But of course this is just a feeling and I don't have any data to back this.

But adding the possibility to have this feature is a good thing I guess, if it seems important enough though 🤔 

One Idea I just have in my mind is that maybe it could be constrained in that if more than for example 3 columns are used, one could create another "block" next to the current block, that also includes up to 3 columns for example that would be visible obvious, that it would wrap on small screen sizes. Meaning we could have blocks (maybe in form of groups?) that can be but next to each other and contain up to 2 or 3 (what ever you think is best) columns, that belong to each other) In this way not making it 100% dynamic and flexible, but still having the ability to put more information next too each other.
Maybe even show a warning explaining the situation and the possible drawbacks for mobile.

**Reply by rijkvanzanten** [reply]
> One Idea I just have in my mind is that maybe it could be constrained in that if more than for example 3 columns are used, one could create another "block" next to the current block, that also includes up to 3 columns for example that would be visible obvious, that it would wrap on small screen sizes. Meaning we could have blocks (maybe in form of groups?) that can be but next to each other and contain up to 2 or 3 (what ever you think is best) columns, that belong to each other) In this way not making it 100% dynamic and flexible, but still having the ability to put more information next too each other.
Maybe even show a warning explaining the situation and the possible drawbacks for mobile.

Yeah something like this feels promising! That way mobile can be 1 column, small can be 3 columns, and desktop+ can be 6

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by w0kyj** [comment]
## Request Review Notes August 29, 2024

- Field label & description layout
    - location of these can affect the rendering and spacing
    - bad design can negatively impact the end user experience
    - 
- Grid Layout
    - responsive design?
    - Form is always one column -- nested groups inside to handle reactive display
        - Could be cumbersome or difficult to maintain
    - Improve and/or simplify the administration of the item layout
        - custom interfaces must be considered
    - add a special component that can be defined for group interfaces, where they decide on the layout for field layout in the data model
- Interface Handling & Display issues
    - Toggle/checkbox -- how should wrapping be handled when many items
    - Floating field groups that wrap/unwrap when the screen is wide enough
    - Mobile & Desktop must be supported
    - 
- Custom Interfaces
    - Need to consider how this will be supported and/or migrated
- UX Design & Compliance
    - Need to ensure a quality end user experience and Directus standards are upheld
    - WCAG -- consider options for ensuring tab ordering (defaults or configurable)
        - review and add support as part of this


### General Decisions
- group design options
- label & description options
    - Allow for group/grid level setting -- top, left positions and left, right, center aligned
- Ensure WCAG accessibility support - related discussions:
    - https://github.com/directus/directus/discussions/8407
    - https://github.com/directus/directus/discussions/4825
    - https://github.com/directus/directus/discussions/21298
    - https://github.com/directus/directus/discussions/2978


Chat Notes:
3/4 columns layout.
Expand form width
Align forms left, right, center.
ACTUAL RELATION DROPDOWNS
Cascading dropdowns
Not 100% form related, but open record on the sidebar directly from the display (e.g. table)
Icon buttons for related M2M fields (I don't need a massive GHASTLY add new or existing record)

**Comment by ralphwest1** [comment]
This would be a really helpful feature. I've submitted it to the new https://roadmap.directus.com with a reference to this discussion. 

Not sure how the new roadmap site works as after submitting my request isn't public and I can only add requests in plain text with no markdown.

If my roadmap request goes public I'll add a link here.

Below is my roadmap request for reference, in case it doesn't become public:

## Allow Admins to Create Multiple Forms (Item Detail Forms) per Collection

Directus currently generates a single Item Detail form for each collection. Its field selection, order, grouping, width, interfaces, and conditions are shared wherever an item is opened.

Please allow administrators to create multiple named Item Detail forms for the same collection. Each form would edit the same underlying items and fields but could provide a different editing experience for a particular role, workflow, or UI entry point.

For example, a `customers` collection might have:

- A comprehensive administrative form.
- A focused sales form.
- A support form.
- A quick-create form containing only essential fields.
- An approval form with relevant fields and read-only context.

### Requested capabilities

- Create multiple named forms for a collection.
- Select a default form.
- Configure the fields, ordering, widths, and groups on each form.
- Control which roles or policies can use each form.
- Open a specific form from a URL, bookmark, dashboard, relation, or workflow.
- Let eligible users switch forms.
- Continue enforcing existing collection and field permissions.

Forms would only control presentation. They would not create separate collections, duplicate data, or replace Directus permissions and validation.

Conditional fields do not completely solve this because they depend on item values rather than the context in which an item was opened. Permissions control security, not form layout. Collection bookmarks currently save list layouts but not Item Detail forms.

### Existing discussion

This request builds upon considerable previous community and maintainer discussion:

- [[Additional Form/Field Layout Options](https://github.com/directus/directus/discussions/9161)](https://github.com/directus/directus/discussions/9161)
- [[Dynamics 365 comparison and multiple-forms suggestion](https://github.com/directus/directus/discussions/9161#discussioncomment-2552427)](https://github.com/directus/directus/discussions/9161#discussioncomment-2552427)
- [[Form bookmarks for different form layouts and purposes](https://github.com/directus/directus/discussions/13423)](https://github.com/directus/directus/discussions/13423)
- [[Review notes considering forms and configurable groups per role](https://github.com/directus/directus/discussions/24919#discussioncomment-12563846)](https://github.com/directus/directus/discussions/24919#discussioncomment-12563846)
